### PR TITLE
Update child_cases_monthly_v2.json

### PR DESCRIPTION
**Pull request by SunainaWalia**
Added new named expressions called zscore_grading_wfa and two new indicators called zscore_grading_wfa_display and resident to resolve the asana  linked below:

https://app.asana.com/0/1173592047649882/1174753167612672

Based on this, I plan to create new sum_when_templates as described in the attached techdoc, and update the UCR V2 MPR 7 Growth monitoring UCR.
https://docs.google.com/document/d/1a9QqCNiJIUeudh06pg-w8050E7hbOSXPJ2XkD0pDOO8/edit#

@mkangia @Rohit25negi, requesting you to review this change.

### CODE REVIEW

File: custom/icds_reports/ucr/data_sources/child_cases_monthly_v2.json [hunk 1]
```diff
@@ -407,6 +407,17 @@
         "type": "boolean",
         "column_id": "child_count_male"
       },
+     {
+      "filter": {
+      "type": "named",
+      "name": "is_resident",
+      "comment": null
```
Comment (mkangia) [inline custom/icds_reports/ucr/data_sources/child_cases_monthly_v2.json:414]: If its not necessary to put `"comment": null`, better to just avoid it since its handled by default. Just blurs the other essential things.

**Review by mkangia — COMMENTED** [review]
@SunainaWalia can you fix indentation please? its a bit hard to read.

Have you tested these changes with a dynamic data source on a test domain?

**Comment by SunainaWalia** [comment]
> @SunainaWalia can you fix indentation please? its a bit hard to read.
> 
> Have you tested these changes with a dynamic data source on a test domain?

Yes, I have tested these in the DS linked below, please feel free to refer the link

https://icds-staging.commcarehq.org/a/icds-cas/configurable_reports/data_sources/edit/a07a277b4e001c33d44182409dd3167e/

**Comment by mkangia** [comment]
> Yes, I have tested these in the DS linked below, please feel free to refer the link

Cool, thanks @SunainaWalia . I did read the conditional one closely but needed more confidence on that.

File: custom/icds_reports/ucr/data_sources/child_cases_monthly_v2.json (hunk 1 above)
Comment (SunainaWalia) [reply custom/icds_reports/ucr/data_sources/child_cases_monthly_v2.json]: have removed comment: null

**Comment by SunainaWalia** [comment]
> > Yes, I have tested these in the DS linked below, please feel free to refer the link
> 
> Cool, thanks @SunainaWalia . I did read the conditional one closely but needed more confidence on that.

I updated indentation as well. Please take a look, thanks!

File: custom/icds_reports/ucr/data_sources/child_cases_monthly_v2.json [hunk 2]
```diff
@@ -2184,8 +2207,68 @@
           "type": "property_name",
           "property_name": "no_exist"
         }
-      }
-    },
```
Comment (mkangia) [inline custom/icds_reports/ucr/data_sources/child_cases_monthly_v2.json:2188]: is this change correct @SunainaWalia ? I think you removed an extra brace here.

Comment (mkangia) [reply custom/icds_reports/ucr/data_sources/child_cases_monthly_v2.json:2188]: oh i think that is because its added in the end. Cool

**Review by mkangia — APPROVED** [review]

**Comment by mkangia** [comment]
@Rohit25negi @calellowitz can one of you take a look as well?

**Comment by SunainaWalia** [comment]
> @Rohit25negi @calellowitz can one of you take a look as well?

Since @Rohit25negi is offline, requesting your feedback here @calellowitz , Thanks!

File: custom/icds_reports/ucr/data_sources/child_cases_monthly_v2.json [hunk 3]
```diff
@@ -1870,7 +1879,21 @@
...
+        "datatype": "string",
+        "type": "expression",
+        "is_primary_key": false,
+        "transform": {},
+        "is_nullable": true,
+        "expression": {
+          "type": "named",
+          "name": "zscore_grading_wfa"
+        },
+        "column_id": "zscore_grading_wfa_display",
```
Comment (calellowitz) [inline custom/icds_reports/ucr/data_sources/child_cases_monthly_v2.json:1894]: this is a weird name for the column, it is probably nicer to just call it `zscore_grading_wfa`, since its not actually displayed and is more concise the other way.

File: custom/icds_reports/ucr/data_sources/child_cases_monthly_v2.json [hunk 4]
```diff
@@ -2184,8 +2207,68 @@
           "type": "property_name",
           "property_name": "no_exist"
         }
-      }
-    },
+      },
+      "zscore_grading_wfa": {
+        "type": "conditional",
```
Comment (calellowitz) [inline custom/icds_reports/ucr/data_sources/child_cases_monthly_v2.json:2212]: this expression does not seem correct. it will overwrite previous data to be `no_exist` once a new gmp date is changed. So the current month will show the zscore, and all previous months will show `no_exist`. Is that what you are intending?

**Review by calellowitz — CHANGES_REQUESTED** [review]
Left one concern about the expression itself, otherwise just going to ask the same thing I ask every time, can this be done in the report instead of in the data source? If any historical data is needed, it is going to take a VERY long time to rebuild this data source.

The ticket looks like it offers an alternative to these monhtly data sources. Why was this option chosen? The monthly data source continues to have consistent lag problems in addition to the problem of rebuilding data above. Furthermore, since this isnt actually a monthly indicator (as per my comment in the change), it can go in the regular (far more efficient) child health data source, although I would recommend staying with form ones if possible, since that is what was used before according to the ticket, and since that is the best way to ensure the data is correct in all months, not just the current one.

File: custom/icds_reports/ucr/data_sources/child_cases_monthly_v2.json (hunk 4 above)
Comment (SunainaWalia) [reply custom/icds_reports/ucr/data_sources/child_cases_monthly_v2.json:2212]: Hi @calellowitz,

Thanks for flagging this, you are right. This expression does not allow us to maintain historic data for the previous month and that is not what is needed. I am exploring an alternative for this.

**Comment by SunainaWalia** [comment]
> Left one concern about the expression itself, otherwise just going to ask the same thing I ask every time, can this be done in the report instead of in the data source? If any historical data is needed, it is going to take a VERY long time to rebuild this data source.
> 
> The ticket looks like it offers an alternative to these monhtly data sources. Why was this option chosen? The monthly data source continues to have consistent lag problems in addition to the problem of rebuilding data above. Furthermore, since this isnt actually a monthly indicator (as per my comment in the change), it can go in the regular (far more efficient) child health data source, although I would recommend staying with form ones if possible, since that is what was used before according to the ticket, and since that is the best way to ensure the data is correct in all months, not just the current one.

Hi @calellowitz 
As discussed offline, for backfilling, we only need to backfill data upto the previous month and not the entire history. I understand that that may be time consuming as well but I reckon it will be easier than backfilling the entire historic data? As far as the question on why monthly data source was chosen is because one, with a form based datasource we are incrementing based on form submission and not based on updates to case. I am not sure if it is even possible to tweak the Form based datasource to increment on case updates and then give an aggregated count of all such cases for the month. Why the non monthly child health DS was not chosen was for the very reason that the data for the previous month will get overwritten when the last_date_gmp is overwritten, which is why the monthly DS seemed like the best option. Also, given now we may not be migrating to the dashboard tables, that option is out of question as well.
At this point I don't know if the form based DS can be utilised. I would like to offline further on this to hear specific ideas from you.

File: custom/icds_reports/ucr/data_sources/child_cases_monthly_v2.json (hunk 3 above)
Comment (SunainaWalia) [reply custom/icds_reports/ucr/data_sources/child_cases_monthly_v2.json:1894]: @calellowitz , can the name of a named_expression and configured_indicator be the same? the _display is to differentiate it from the named_expression.

**Comment by SunainaWalia** [comment]
Update on this PR : I had an offline discussion with @calellowitz  and based on the discussion am exploring making this change in the form based data source itself instead of editing this monthly data source. So this PR is stale and I will be raising a separate PR once the change has been designed, implemented and tested. Let me know if you have any further questions here or if we should close this PR @mkangia

**Comment by mkangia** [comment]
Can you summarize the reason behind that decision and close it? @SunainaWalia

**Comment by SunainaWalia** [comment]
> Can you summarize the reason behind that decision and close it? @SunainaWalia

Sure. We should close this PR as 
1. This change only partially addresses what we intend to do. It is only allowing us to fetch results for current/latest changes made to the case but wipes out the changes for the past months, which needs to be corrected.
2. Monthly data sources as it is are slow and have processing lags; adding more complex expressions will further degrade the quality of the data source and introduce more burden in terms of lags, processing speeds etc. which is expensive from a system point of view. So I will be exploring making this change using a growth monitoring forms data source. @mkangia

**Comment by mkangia** [comment]
thank you @SunainaWalia . that makes sense.
